server/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -69,35 +68,61 @@ static LogicalPlan buildJoinTree(List<AnalyzedRelation> from,
         Map<Set<RelationName>, Symbol> queryParts = QuerySplitter.split(whereClause);
         List<JoinPair> allJoinPairs = convertImplicitJoinConditionsToJoinPairs(joinPairs, queryParts);
         LinkedHashMap<Set<RelationName>, JoinPair> joinPairsByRelations = buildRelationsToJoinPairsMap(allJoinPairs);
-        Iterator<RelationName> it = Lists.mapLazy(from, AnalyzedRelation::relationName).iterator();
 
-        final RelationName lhsName = it.next();
-        final RelationName rhsName = it.next();
-        Set<RelationName> joinNames = new LinkedHashSet<>();
-        joinNames.add(lhsName);
-        joinNames.add(rhsName);
+        Map<RelationName, AnalyzedRelation> relationsFromClause = new HashMap<>();
+        List<RelationName> relationNamesFromClause = new ArrayList<>();
+
+        for (AnalyzedRelation analyzedRelation : from) {
+            RelationName relationName = analyzedRelation.relationName();
+            relationNamesFromClause.add(relationName);
+            relationsFromClause.put(relationName, analyzedRelation);
+        }
+
+        final RelationName lhsName;
+        final RelationName rhsName;
 
-        JoinPair joinLhsRhs = joinPairsByRelations.remove(joinNames);
         final JoinType joinType;
         final Symbol joinCondition;
-        if (joinLhsRhs == null) {
-            joinType = JoinType.CROSS;
-            joinCondition = null;
+        Set<RelationName> joinNames = new LinkedHashSet<>();
+
+        // Outer-joins must be created before the implicit joins because the are
+        // binding stronger since they are order-dependent.
+        if (containsOuterJoins(allJoinPairs)) {
+            JoinPair joinPair = allJoinPairs.removeFirst();
+            lhsName = joinPair.left();
+            rhsName = joinPair.right();
```

Comment:
AFAIK, `allJoinPairs` doesn't guarantee that outer join pairs come first. 

Do I understand it correctly that we basically pick up arbitrary pair here?

Reply:
We start with the explicit joins in the order of the query.